BLE Indoor Positioning/src/test/java/com/nexenio/bleindoorpositioning/location/provider/LocationPredictorTest.java

```diff
@@ -0,0 +1,102 @@
+        double delta = predictedLocation.getDistanceTo(FIFTH_POINT_OF_LINE);
+        assertEquals(0, delta, 5);
+    }
+
+    @Test
+    public void predictLocationFromLocations_locationsInCircle_accuratePrediction() throws Exception {
+        long timestampDelta = 500;
+        double distanceToNextLocation = HUMAN_WALKING_SPEED / TimeUnit.MILLISECONDS.toSeconds(timestampDelta);
+        List<Location> walkInACircle = new ArrayList<>();
+        Location nextLocation;
+        Location lastLocation = null;
+        for (int angle = 0; angle < 360; angle += 10) {
+            if (angle == 0) {
+                nextLocation = new Location(FIRST_POINT_OF_LINE).getShiftedLocation(distanceToNextLocation, angle);
+            } else {
+                nextLocation = lastLocation.getShiftedLocation(distanceToNextLocation, angle);
+                nextLocation.setTimestamp(lastLocation.getTimestamp() + timestampDelta);
+            }
+            walkInACircle.add(nextLocation);
+            lastLocation = nextLocation;
+        }
+
+        for (int i = 0; i < walkInACircle.size() - 3; i++) {
+            List<Location> subList = new ArrayList<>();
+            subList.add(walkInACircle.get(i));
+            subList.add(walkInACircle.get(i + 1));
+
+            Location lastKnownLocation = walkInACircle.get(i + 2);
+            subList.add(lastKnownLocation);
+
+            Location expectedLocation = walkInACircle.get(i + 3);
+            Location predictedLocation = LocationPredictor.predict(subList, TimeUnit.SECONDS.toMillis(1));
+
+            // angle
+            double expectedAngle = lastKnownLocation.getAngleTo(expectedLocation);
+            double actualAngle = lastKnownLocation.getAngleTo(predictedLocation);
+            assertEquals(expectedAngle, actualAngle, 50);
+
+            // distance
+            double expectedDistance = lastKnownLocation.getDistanceTo(expectedLocation);
+            double actualDistance = lastKnownLocation.getDistanceTo(predictedLocation);
+            assertEquals(expectedDistance, actualDistance, 1);
+        }
+
+    }
+
+    @Test
+    public void calculateSpeed_locations_correctSpeed() throws Exception {
+        List<Location> lineOnGendarmenmarkt = new ArrayList<>(Arrays.asList(
+                new Location(FIRST_POINT_OF_LINE), new Location(SECOND_POINT_OF_LINE),
+                new Location(THIRD_POINT_OF_LINE), new Location(FOURTH_POINT_OF_LINE)
+        ));
+        lineOnGendarmenmarkt = setupTimestampsForLocations(lineOnGendarmenmarkt);
+        double speed = LocationPredictor.calculateSpeed(lineOnGendarmenmarkt);
+        assertEquals(speed, HUMAN_WALKING_SPEED, 0.1);
+    }
+
+    private static List<Location> setupTimestampsForLocations(List<Location> locations) {
+        for (int i = 1; i < locations.size(); i++) {
+            // set timestamps based on timestamp of previous square and human walking speed
```

Comment:
We have no square.

Reply:
I was referring to a square on the floor, which was my variable name for locations earlier